Replace git-based Bioconductor R packages

codeandkey opened the pull request:
The current collection of Bioconductor packages tend to have scattered
dependencies and missing versions. Many have not been updated to the newest release making them incompatible with R >= 3.6. This commit replaces git-based
packages with tool-generated Spack package recipes with correct
dependencies and descriptions in place.

The new packages here have all available versions summed from Bioconductor releases 3.5, 3.6, 3.7, 3.8, and 3.9.

The generated packages do not contain any extra required logic for building the package; this will require manual review and testing.
Most non-git based Bioconductor packages (EG r-genomeinfodbdata) and some other CRAN packages will need to be updated before this is merged.

I'm currently working on installing each package to find any problems.

codeandkey commented:
I can't yet add labels but this PR is still a work in progress.

codeandkey commented:
Looks like none of the updated packages used any extra `configure_args` so that should be OK. Some newer versions brought in some new dependency packages which were generated in the same manner. Some updates to CRAN packages were necessary (`r-clue`, `r-vgam`) which may require merging with the pending updates in #12019.

codeandkey commented:
I've gone through each package and made sure they concretize properly. I have yet to install each one yet though.

The updates indirectly bring in the following dependencies as new packages.

CRAN:
- r-clue
- r-scrime

Bioconductor:
- r-gofuncr
- r-rnaseqmap
- r-rhdf5lib
- r-delayedmatrixstats
- r-hdf5array
- r-biocfilecache
- r-genemeta

codeandkey commented:
Is there an official spack stance on including [BiocManager](https://cran.r-project.org/web/packages/BiocManager/index.html)? It's a tool for installing packages so it doesn't fit very well into the spack ecosystem but it is required as a dependency for many packages in Bioconductor.

baberlevi commented:
> Is there an official stance on including [BiocManager](https://cran.r-project.org/web/packages/BiocManager/index.html)? It's a tool for installing packages so it doesn't fit very well into the spack ecosystem but it is required as a dependency for many packages in Bioconductor.

I think we'll have to include it, and we have other similar tools already (e.g. pip)

codeandkey commented:
Each bio package has concretized and built successfully. I think this is about ready for another pair of eyes.

#12019 should be merged before this one. This PR includes many necessary updates to CRAN packages which will be made obsolete.

adamjstewart commented:
#12019 has now been merged. Can you rebase? Is this ready to review?

One of the main reasons for #12019 was to allow #12015. I'm not sure how many packages are from neither CRAN nor Bioconductor, but can you make sure all of the non-CRAN packages have all of their dependencies properly declared so we can merge #12015? I'm going to make this PR the non-CRAN PR if you don't mind.

codeandkey commented:
That's fine. I will work on rebasing this PR to reflect the changes merged in #12019.

adamjstewart commented:
@codeandkey is this ready to review yet?

codeandkey commented:
Yes this should be ready. I haven't tested the packages after the rebase but it doesn't look like any bio packages were changed. All the overlapping CRAN packages looked OK too -- I overwrote them with the new files from #12019.

codeandkey commented:
Yes, the descriptions are combined in a script. I'll add some more criterion for concatenating them.
As for the R versions, those follow how Bioconductor packages were originally written. The R versions are matched to each Bioconductor release [here](https://bioconductor.org/about/release-announcements/).
It would be quick to have the R constraints follow the actual dependencies from the DESCRIPTION if that would be better.

adamjstewart commented:
For the R version dependencies, I don't actually use R, so I'm hoping the R users in Spack can decide what would be best.

glennpj commented:
The Bioconductor stuff is a pain to try to map to spack. Not all packages will list version ranges but the bioconductor installer will pick up the appropriate version for the bioconductor/R release being installed into. In other words, I do not think the DESCRIPTION files for bioconductor packages are sufficient to describe the dependencies.

Also, if specifying the R version to correspond to a package of a given bioconductor release then it is also necessary to do the same for the package dependencies. For example:

```
    depends_on('r@3.6.0:3.6.9', when='@1.32.0', type=('build', 'run'))
    depends_on('r@3.5.0:3.5.9', when='@1.30.0', type=('build', 'run'))
    depends_on('r@3.5.0:3.5.9', when='@1.28.0', type=('build', 'run'))
    depends_on('r@3.4.0:3.4.9', when='@1.26.0', type=('build', 'run'))
    depends_on('r@3.4.0:3.4.9', when='@1.24.0', type=('build', 'run'))

    depends_on('r-a4base', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4classif', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4core', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4preproc', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4reporting', when='@1.24.0:', type=('build', 'run'))
```
That does not specify the restraints on the a4* packages. For instance, r-a4base would need to be 
```
    depends_on('r-a4base@1.24.0', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4base@1.26.0', when='@1.26.0:', type=('build', 'run'))
    depends_on('r-a4base@1.28.0', when='@1.28.0:', type=('build', 'run'))
    depends_on('r-a4base@1.30.0', when='@1.30.0:', type=('build', 'run'))
    depends_on('r-a4base@1.32.0', when='@1.32.0:', type=('build', 'run'))
```
and likewise for r-a4classif, r-a4core, r-a4preproc, and r-a4reporting.

glennpj commented:
> In other words, I do not think the DESCRIPTION files for bioconductor packages are sufficient to describe the dependencies.
To follow up on that, the Bioconductor installer will use metadata found on the repository site in combination with the local knowledge of the R and Bioconductor releases to determine which version of any given package to install. I do not know if there is a way to pull that information to have it available for spack.

adamjstewart commented:
> Also, if specifying the R version to correspond to a package of a given bioconductor release then it is also necessary to do the same for the package dependencies. For example, the r-a4 package has the following:

I think that's true for the current buggy concretizer, but hopefully shouldn't be a requirement for the future concretizer?

codeandkey commented:
Thanks very much @glennpj for your input here. Looks like we are left with a choice to try and keep bioconductor releases bundled together, or to rely on the DESCRIPTION for the actual dependencies. Personally I have not personally had to work with many installations of bioconductor packages so I'm not familiar with the advantages of staying within a particular release.

A crazy thought: we could use variants in an interesting way to enforce bioconductor releases. For instance, we could add a 'b3.5' variant which enforces an extra set of dependencies with versions matching those in the 3.5 release, each with the 'b3.5' variant as well. The `r-a4` package would then look something like this:
```
    variant('b3.5', description='Enforce Bioconductor release 3.5', default=False)

    depends_on('r@3.4.0:3.4.9', when='+b3.5', type=('build', 'run'))

    depends_on('r-a4base', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4classif', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4core', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4preproc', when='@1.24.0:', type=('build', 'run'))
    depends_on('r-a4reporting', when='@1.24.0:', type=('build', 'run'))

    depends_on('r-a4base+b3.5@1.24.0', when='+b3.5', type=('build', 'run'))
    # ... and likewise for each dependency and each bioconductor release checksummed
```
This method could make package files much larger if we want to have a lot of releases available but perhaps we could just checksum a couple instead of 5.

adamjstewart commented:
Sorry for being MIA, I just dropped all of my courses this semester so I have more free time to review PRs now.

Have we reached a consensus on how best to handle Bioconductor releases and R version dependencies? Pinging @HenrikBengtsson who has helped a lot with my questions on R in the past.

glennpj commented:
I did not realize that there was an open question regarding the proposed variant scheme, so I apologize for not commenting sooner. Variants do not propagate, as far as I know, so I think you would still have to do all of the version specifications that I outlined earlier even if you specified bioconductor releases as variants. I do wonder if BundlePackage could be leveraged for that.

adamjstewart commented:
Personally, I'm leaning towards relying on the R version requirements specified in DESCRIPTION and adding a BundlePackage that specifies which version of each bioconductor package is required by each release.

adamjstewart commented:
Ping @codeandkey. If no one else has any objections, let's go with my previous suggestion. I would love to get this PR and #12015 merged before the next Spack release.

codeandkey commented:
I think BundlePackage will work well for this. I'm working on rewriting the tool to grab all dependencies from the DESCRIPTION.

codeandkey commented:
The packages have been updated to use the DESCRIPTION for all dependencies, so many of the packages no longer have an R version constraint at all. The RPackage class depends on R unconditionally so I don't think this will be an issue. I'm not sure how to go about making the BundlePackage for the bioconductor releases however; perhaps it would be better in another PR?

adamjstewart commented:
I'm fine with pushing the BundlePackage to a separate PR. It should be pretty easy to do.

Can you resolve the merge conflicts?

codeandkey commented:
Merge should be ready, but looks like my tool missed a couple more packages. I'm going to quickly walk through them and catch those.

codeandkey commented:
This should be good to go now.

adamjstewart commented:
I'll try to review this over the weekend.

adamjstewart commented:
One of your packages uses md5 and needs to be updated to sha256. This is causing the unit tests to fail.

adamjstewart commented on var/spack/repos/builtin/packages/r-a4/package.py:
```diff
@@ -7,16 +7,21 @@
+    """Automated Affymetrix Array Analysis Umbrella Package.
 
-    homepage = "https://www.bioconductor.org/packages/a4/"
+       Automated Affymetrix Array Analysis Umbrella Package"""
```
This description is duplicated. I assume this came from combining the short and long descriptions? I would either only use one of them, or check if the long description is the same (or shorter) than the short description.

codeandkey replied:
Updated to follow this criteria and it seems to have cleaned up most of the redundant descriptions.

adamjstewart commented on var/spack/repos/builtin/packages/r-a4/package.py:
```diff
@@ -7,16 +7,21 @@
+    depends_on('r-a4classif', when='@1.24.0:', type=('build', 'run'))
+    depends_on('r-a4core', when='@1.24.0:', type=('build', 'run'))
+    depends_on('r-a4preproc', when='@1.24.0:', type=('build', 'run'))
+    depends_on('r-a4reporting', when='@1.24.0:', type=('build', 'run'))
```
Unless you have some reason to believe that version 1.22.0 did not require these dependencies, I would drop the `when` from the package.

codeandkey replied:
Updated. Dropped the 'when' for all of the oldest checksummed versions.

adamjstewart commented on var/spack/repos/builtin/packages/r-abadata/package.py:
```diff
@@ -7,16 +7,24 @@
+       data from the Developing Human Brain pooled into five age categories and
+       averaged across donors and (3) a developmental effect score based on the
+       Developing Human Brain expression data. All datasets are restricted to
+       protein coding genes."""
```
This is an example of a package that actually benefits from combining the short and long description, although I think the short description is sufficient if you want to simplify things that way.

adamjstewart commented on var/spack/repos/builtin/packages/r-abadata/package.py:
```diff
@@ -7,16 +7,24 @@
+    version('1.8.0', commit='181a4af1af349064eb432255970e925ae2564e1a')
+    version('1.6.0', commit='517c18a3d1809dde0291eeb47dd2545c7cfcdabe')
+
+    depends_on('r@3.2:', when='@1.6.0:', type=('build', 'run'))
```
Same here, `when` not needed unless you know when this version requirement was added.

adamjstewart commented on var/spack/repos/builtin/packages/r-acde/package.py:
```diff
@@ -7,21 +7,27 @@
+    version('1.6.0', commit='244c81f435a077bf7895ea565fa6695e8b079f67')
+
+    depends_on('r-boot@1.3:', when='@1.6.0:', type=('build', 'run'))
+    depends_on('r@3.3:', when='@1.6.0:', type=('build', 'run'))
```
I prefer to order dependencies in the same order they appear in the description file, just so it's easier to check if they are all present. I also put R first in the list. Not required, but would make this easier to review if it isn't much work to change in your script.

codeandkey replied:
Updated -- the dependencies are now ordered as they appear in the description.

adamjstewart commented on var/spack/repos/builtin/packages/r-aneufinder/package.py:
```diff
@@ -33,6 +40,10 @@ class RAneufinder(RPackage):
     depends_on('r-ggplot2', type=('build', 'run'))
```
It doesn't look like `r-biocgenerics` is a dependency.

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-annotationforge/package.py:
```diff
@@ -7,19 +7,24 @@
-    depends_on('r-biocgenerics', type=('build', 'run'))
-    depends_on('r-biobase', type=('build', 'run'))
-    depends_on('r-annotationdbi', type=('build', 'run'))
+    depends_on('r@2.7.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r@3.5.0:', type=('build', 'run'))
```

codeandkey replied:
This is very strange. The Bioconductor website shows a minimum R version of 3.5.0 while the DESCRIPTION in the repository has a minimum version of 2.7.0.

adamjstewart replied:
Still not sure what we want to do here, I'll let you decide.

adamjstewart replied:
Let's trust the repository, that's what will actually crash if you try to install with the wrong version of R

adamjstewart commented on var/spack/repos/builtin/packages/r-beachmat/package.py:
```diff
@@ -0,0 +1,40 @@
+
+    depends_on('r-hdf5array@1.9.5:', when='@1.4.0:', type=('build', 'run'))
+    depends_on('r-delayedarray@0.7.38:', when='@1.4.0:', type=('build', 'run'))
+    depends_on('r-biocgenerics', when='@1.4.0:', type=('build', 'run'))
```
The current version of this package doesn't seem to have most of these dependencies.

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-biocinstaller/package.py:
```diff
@@ -7,12 +7,20 @@
+       This package is used to install and update Bioconductor, CRAN, and
+       (some) github packages."""
+
+    homepage = "https://bioconductor.org/packages/BiocInstaller"
```
```suggestion
    homepage = "https://bioconductor.org/packages/3.8/bioc/html/BiocInstaller.html"
```

adamjstewart replied:
Ping

adamjstewart commented on var/spack/repos/builtin/packages/r-biocinstaller/package.py:
```diff
@@ -7,12 +7,20 @@
-    depends_on('r@3.4.0:3.4.9', when='@1.26.1')
+    depends_on('r@3.4.0:', type=('build', 'run'))
+
+    depends_on('r@3.5.0:', when='@1.30.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r@3.5.0:3.5.999', when='@1.30.0:', type=('build', 'run'))
```
Not sure about previous releases.

adamjstewart replied:
Ping

adamjstewart replied:
Checked the git repo and it looks like this isn't the case in the DESCRIPTION file.

adamjstewart commented on var/spack/repos/builtin/packages/r-biocmanager/package.py:
```diff
@@ -0,0 +1,18 @@
+
+    version('1.30.4', sha256='50093f5c8ed8fba6e68bc715784b713887bdad3538fbb92f152dcc1eaf39ba4f')
+
+    depends_on('r@3.5.0:', type=('build', 'run'))
```
I don't see this restriction.

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-biocneighbors/package.py:
```diff
@@ -0,0 +1,35 @@
+    version('1.2.0', commit='f754c6300f835142536a4594ddf750481e0fe273')
+    version('1.0.0', commit='e252fc04b6d22097f2c5f74406e77d85e7060770')
+
+    depends_on('r@3.5:', type=('build', 'run'))
```
I don't see this restriction

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented:
Can you rebase `r-mzr`? I renamed `netcdf` to `netcdf-c` in #13411.

adamjstewart commented on var/spack/repos/builtin/packages/r-dose/package.py:
```diff
@@ -7,27 +7,37 @@
+    depends_on('r-s4vectors', type=('build', 'run'))
+    depends_on('r-scales', type=('build', 'run'))
+
+    depends_on('r-rvcheck', when='@3.4.0:', type=('build', 'run'))
```
Not required, at least in latest version.

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-ensembldb/package.py:
```diff
@@ -7,30 +7,39 @@
+    depends_on('r-annotationdbi@1.31.19:', type=('build', 'run'))
     depends_on('r-rtracklayer', type=('build', 'run'))
     depends_on('r-s4vectors', type=('build', 'run'))
     depends_on('r-annotationhub', type=('build', 'run'))
```
This is no longer required

adamjstewart replied:
Ping

codeandkey replied:
fixed!

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-genomeinfodbdata/package.py:
```diff
@@ -13,8 +13,12 @@ class RGenomeinfodbdata(RPackage):
+
     depends_on('r@3.4.0:3.4.9', when='@0.99.0', type=('build', 'run'))
     depends_on('r@3.5.0:3.5.9', when='@1.1.0', type=('build', 'run'))
+    depends_on('r@3.6.0:3.6.9', when='@1.2.1', type=('build', 'run'))
```
It looks like this package wasn't updated to use the new git checkout and R versions

codeandkey replied:
GenomeInfoDbData doesn't have a remote git for some reason. It is only accessible via the tarball -- I'll update it to follow the same dependency pattern as the others.

adamjstewart replied:
That's so weird...

adamjstewart replied:
Ping, still needs to be updated I think.

codeandkey replied:
Fixed!

adamjstewart commented on var/spack/repos/builtin/packages/r-genomicfeatures/package.py:
```diff
@@ -16,35 +18,40 @@ class RGenomicfeatures(RPackage):
+    depends_on('r-iranges@2.11.16:', when='@1.30.3:', type=('build', 'run'))
+    depends_on('r-genomeinfodb@1.13.1:', when='@1.30.3:', type=('build', 'run'))
+    depends_on('r-genomicranges@1.29.14:', when='@1.30.3:', type=('build', 'run'))
+    depends_on('r-rmysql', when='@1.30.3:', type=('build', 'run'))
```
Not required in latest version

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-geoquery/package.py:
```diff
@@ -7,18 +7,32 @@
     version('2.42.0', commit='c26adef8d3ddbd6932a3170f2f84f6e4327641fb')
 
     depends_on('r-biobase', type=('build', 'run'))
     depends_on('r-xml', type=('build', 'run'))
```
No longer required

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-geoquery/package.py:
```diff
@@ -7,18 +7,32 @@
 
     depends_on('r-biobase', type=('build', 'run'))
     depends_on('r-xml', type=('build', 'run'))
     depends_on('r-rcurl', type=('build', 'run'))
```
No longer required

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-iranges/package.py:
```diff
@@ -7,29 +7,34 @@
+
+    depends_on('r-s4vectors@0.19.11:', when='@2.16.0:', type=('build', 'run'))
+
+    depends_on('r-s4vectors@0.21.9:', when='@2.18.3:', type=('build', 'run'))
```
A new version was just released, bumps the s4vectors version requirement. You can add it if you want to, but certainly not required.

adamjstewart commented on var/spack/repos/builtin/packages/r-mzr/package.py:
```diff
@@ -7,22 +7,33 @@
+
+    depends_on('r-rhdf5lib@1.1.4:', when='@2.14.0:', type=('build', 'run'))
+
+    depends_on('r-ncdf4', when='@2.16.2:', type=('build', 'run'))
```
Missing build dep on gmake

adamjstewart replied:
Still missing a build dep on gmake

adamjstewart commented on var/spack/repos/builtin/packages/r-oligoclasses/package.py:
```diff
@@ -7,25 +7,36 @@
+    depends_on('r-affyio@1.23.2:', type=('build', 'run'))
     depends_on('r-ff', type=('build', 'run'))
     depends_on('r-foreach', type=('build', 'run'))
     depends_on('r-biocinstaller', type=('build', 'run'))
```
I think this dep was replaced with `r-biocmanager`, so it is no longer needed in newer versions.

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-org-hs-eg-db/package.py:
```diff
@@ -13,7 +13,12 @@ class ROrgHsEgDb(RPackage):
     version('3.4.1', sha256='0f87b3f1925a1d7007e5ad9200bdf511788bd1d7cb76f1121feeb109889c2b00')
 
+    depends_on('r@3.6.0:3.6.9', when='@3.8.2')
     depends_on('r@3.4.0:3.4.9', when='@3.4.1')
```
This package still needs to be updated.

codeandkey replied:
Updated

adamjstewart commented on var/spack/repos/builtin/packages/r-organismdbi/package.py:
```diff
@@ -7,24 +7,35 @@
+    depends_on('r-annotationdbi@1.33.15:', type=('build', 'run'))
+    depends_on('r-genomicfeatures@1.23.31:', type=('build', 'run'))
     depends_on('r-biobase', type=('build', 'run'))
     depends_on('r-biocinstaller', type=('build', 'run'))
```
Replaced by biocmanager, no longer needed in newer releases

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-pcamethods/package.py:
```diff
@@ -7,23 +7,28 @@
-    depends_on('r-rcpp', type=('build', 'run'))
+    depends_on('r-rcpp@0.11.3:', type=('build', 'run'))
     depends_on('r-mass', type=('build', 'run'))
-    depends_on('r@3.4.0:3.4.9', when='@1.68.0')
```
`rcpp` is also listed as a system requirement. I assume this is a typo?

adamjstewart commented on var/spack/repos/builtin/packages/r-pfam-db/package.py:
```diff
@@ -13,7 +13,12 @@ class RPfamDb(RPackage):
     version('3.4.1', sha256='fc45a0d53139daf85873f67bd3f1b68f2d883617f4447caddbd2d7dcc58a393f')
 
     depends_on('r@3.4.0:3.4.9', when='@3.4.1')
+    depends_on('r@3.6.0:3.6.0', when='@3.8.2')
```
This package still needs to be updated

codeandkey replied:
PFAM.db is another one that seems to be missing a git repository. Updating manually

adamjstewart commented on var/spack/repos/builtin/packages/r-rgraphviz/package.py:
```diff
@@ -7,13 +7,19 @@
 
-    depends_on('r@3.4.0:3.4.9', when='@2.20.0')
+    depends_on('r@2.6.0:', type=('build', 'run'))
     depends_on('r-graph', type=('build', 'run'))
```
This system requirement is probably worth adding:

> optionally Graphviz (>= 2.16)

codeandkey commented on var/spack/repos/builtin/packages/r-dirichletmultinomial/package.py:
```diff
@@ -7,19 +7,24 @@
 
     depends_on('r-s4vectors', type=('build', 'run'))
     depends_on('r-iranges', type=('build', 'run'))
-    depends_on('gsl')
```
Added

adamjstewart commented on var/spack/repos/builtin/packages/r-rhdf5/package.py:
```diff
@@ -7,18 +7,28 @@
+    depends_on('r-rhdf5lib', when='@2.24.0:', type=('build', 'run'))
+
+    depends_on('r@3.5.0:', when='@2.26.2:', type=('build', 'run'))
+    depends_on('r-rhdf5lib@1.3.2:', when='@2.26.2:', type=('build', 'run'))
```
Missing build dep on gmake

adamjstewart commented on var/spack/repos/builtin/packages/r-rhdf5lib/package.py:
```diff
@@ -0,0 +1,20 @@
+    version('1.6.3', commit='11ac3069f28b0068002db9c645817401f6c5b3c4')
+    version('1.4.3', commit='f6be8c2659b2daa17541506058917b7981490d65')
+    version('1.2.1', commit='dbf85dbedb736d5a696794d52875729c8514494e')
+    version('1.0.0', commit='79608038c2016a518ba747fe6a2bf02ce53a75f9')
```
Missing build dep on gmake

adamjstewart replied:
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-rhtslib/package.py:
```diff
@@ -7,17 +7,21 @@
 
-    depends_on('r@3.4.0:3.4.9', when='@1.8.0')
     depends_on('r-zlibbioc', type=('build', 'run'))
-    depends_on('autoconf@2.67:', type='build')
```
Missing system requirements:

> libbz2 & liblzma & libcurl (with header files), GNU make

adamjstewart commented on var/spack/repos/builtin/packages/r-rsamtools/package.py:
```diff
@@ -7,33 +7,39 @@
+    depends_on('r-iranges@2.13.12:', when='@1.32.3:', type=('build', 'run'))
+    depends_on('r-xvector@0.19.7:', when='@1.32.3:', type=('build', 'run'))
+
+    depends_on('r-rhtslib@1.16.3:', when='@2.0.3:', type=('build', 'run'))
```
Missing build dep on gmake

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    version('1.4.0', commit='90a2eab66ff82ba6dd7fbb33e41cd0ded20fa218')
+
+    depends_on('r@3.3:', type=('build', 'run'))
+    depends_on('r-biobase', type=('build', 'run'))
```
No longer required in latest version.

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r@3.3:', type=('build', 'run'))
+    depends_on('r-biobase', type=('build', 'run'))
+    depends_on('r-ggplot2', type=('build', 'run'))
+    depends_on('r-biomart', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-ggplot2', type=('build', 'run'))
+    depends_on('r-biomart', type=('build', 'run'))
+    depends_on('r-biocgenerics', type=('build', 'run'))
+    depends_on('r-data-table', type=('build', 'run'))
```
No longer required

codeandkey commented on var/spack/repos/builtin/packages/r-ggraph/package.py:
```diff
@@ -0,0 +1,21 @@
+
+    depends_on('r@2.10:', type=('build', 'run'))
+    depends_on('r-ggplot2@2.0.0:', type=('build', 'run'))
+    depends_on('r-rcpp', type=('build', 'run'))
```
Updated, the dependencies have brought in 4 more packages from CRAN: `r-ggforce`, `r-polyclip`, `r-farver`, `r-tweenr`

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-biomart', type=('build', 'run'))
+    depends_on('r-biocgenerics', type=('build', 'run'))
+    depends_on('r-data-table', type=('build', 'run'))
+    depends_on('r-dplyr', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-biocgenerics', type=('build', 'run'))
+    depends_on('r-data-table', type=('build', 'run'))
+    depends_on('r-dplyr', type=('build', 'run'))
+    depends_on('r-edger', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-dplyr', type=('build', 'run'))
+    depends_on('r-edger', type=('build', 'run'))
+    depends_on('r-ggbeeswarm', type=('build', 'run'))
+    depends_on('r-limma', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-ggbeeswarm', type=('build', 'run'))
+    depends_on('r-limma', type=('build', 'run'))
+    depends_on('r-matrix', type=('build', 'run'))
+    depends_on('r-matrixstats', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-limma', type=('build', 'run'))
+    depends_on('r-matrix', type=('build', 'run'))
+    depends_on('r-matrixstats', type=('build', 'run'))
+    depends_on('r-plyr', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on('r-matrix', type=('build', 'run'))
+    depends_on('r-matrixstats', type=('build', 'run'))
+    depends_on('r-plyr', type=('build', 'run'))
+    depends_on('r-reshape2', type=('build', 'run'))
```
No longer required

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+
+    depends_on('r@3.6:', when='@1.12.2:', type=('build', 'run'))
+    depends_on('r-biocneighbors', when='@1.12.2:', type=('build', 'run'))
+    depends_on('r-biocsingular', when='@1.12.2:', type=('build', 'run'))
```
I gave up on this package, let me know when it's accurate and I'll review again lol

adamjstewart commented on var/spack/repos/builtin/packages/r-farver/package.py:
```diff
@@ -0,0 +1,19 @@
+    version('1.1.0', sha256='2086f309135f37705280fe2df851ad91dc886ad8f2a6eb1f3983aa20427f94b6')
+
+    depends_on('r-rcpp@0.12.15:', type=('build', 'run'))
+    depends_on('r-rcpp', type=('build', 'run'))
```
Remove this line

adamjstewart commented on var/spack/repos/builtin/packages/r-tweenr/package.py:
```diff
@@ -0,0 +1,23 @@
+    depends_on('r-farver', type=('build', 'run'))
+    depends_on('r-magrittr', type=('build', 'run'))
+    depends_on('r-rcpp@0.12.3:', type=('build', 'run'))
+    depends_on('r-rcpp', type=('build', 'run'))
```
Delete this line

adamjstewart commented on var/spack/repos/builtin/packages/r-variantannotation/package.py:
```diff
@@ -7,28 +7,51 @@
+    depends_on('r-rsamtools@1.33.6:', when='@1.28.13:', type=('build', 'run'))
+
+    depends_on('r-rsamtools@1.99.0:', when='@1.30.1:', type=('build', 'run'))
+    depends_on('r-rhtslib', when='@1.30.1:', type=('build', 'run'))
```
Missing build dep on gmake

adamjstewart commented on var/spack/repos/builtin/packages/r-vsn/package.py:
```diff
@@ -7,28 +7,36 @@
     depends_on('r-limma', type=('build', 'run'))
     depends_on('r-lattice', type=('build', 'run'))
     depends_on('r-ggplot2', type=('build', 'run'))
     depends_on('r-hexbin', type=('build', 'run'))
```
Looks like this is no longer required?

codeandkey replied:
fixed

adamjstewart commented:
Ping @codeandkey

codeandkey commented:
Sorry for the delay. The script did not account for dependencies made obsolete in later versions so I'm going through the packages where this happens and determining when the dependencies are dropped.

ifelsefi commented:
Any ETA on when this can be merged?

codeandkey commented:
Ping @adamjstewart, this should be ready for another look.

adamjstewart commented:
I'll review this tonight or tomorrow!

adamjstewart commented on var/spack/repos/builtin/packages/r-reportingtools/package.py:
```diff
@@ -7,42 +7,46 @@
     depends_on('r-xml', type=('build', 'run'))
     depends_on('r-r-utils', type=('build', 'run'))
-    depends_on('r-deseq2', type=('build', 'run'))
+    depends_on('r-utils', type=('build', 'run'))
```
This is a builtin library, you can remove this dependency.

codeandkey replied:
Fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-farver/package.py:
```diff
@@ -0,0 +1,18 @@
+
+    version('1.1.0', sha256='2086f309135f37705280fe2df851ad91dc886ad8f2a6eb1f3983aa20427f94b6')
+
+    depends_on('r-rcpp@0.12.15:', type=('build', 'run'))
```
I don't see this dependency on CRAN. Was it removed?

codeandkey replied:
Looks like it was, I just replaced the recipe with one for the newer version.

adamjstewart commented on var/spack/repos/builtin/packages/r-mzr/package.py:
```diff
@@ -7,22 +7,33 @@
+    depends_on('r-rhdf5lib@1.1.4:', when='@2.14.0:', type=('build', 'run'))
+
+    depends_on('r-ncdf4', when='@2.16.2:', type=('build', 'run'))
     depends_on('netcdf-c')
```
This dependency should be removed

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-rhdf5/package.py:
```diff
@@ -7,18 +7,30 @@
+    depends_on('r@3.5.0:', when='@2.26.2:', type=('build', 'run'))
+    depends_on('r-rhdf5lib@1.3.2:', when='@2.26.2:', type=('build', 'run'))
+
+    depends_on('gmake')
```
```suggestion
    depends_on('gmake', type='build')
```

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-rhtslib/package.py:
```diff
@@ -7,17 +7,26 @@
-    depends_on('autoconf@2.67:', type='build')
+
+    depends_on('bzip2', type=('build', 'run'))
+    depends_on('lzma', type=('build', 'run'))
```
lzma is deprecated, use xz instead:
```suggestion
    depends_on('xz', type=('build', 'run'))
```

codeandkey replied:
done!

adamjstewart commented on var/spack/repos/builtin/packages/r-rhtslib/package.py:
```diff
@@ -7,17 +7,26 @@
+    depends_on('bzip2', type=('build', 'run'))
+    depends_on('lzma', type=('build', 'run'))
+    depends_on('curl', type=('build', 'run'))
+    depends_on('gmake', type=('build', 'run'))
```
```suggestion
    depends_on('gmake', type='build')
```

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-rsamtools/package.py:
```diff
@@ -7,33 +7,41 @@
+
+    depends_on('r-rhtslib@1.16.3:', when='@2.0.3:', type=('build', 'run'))
+
+    depends_on('gmake')
```
```suggestion
    depends_on('gmake', type='build')
```

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-scater/package.py:
```diff
@@ -0,0 +1,63 @@
+    version('1.6.3', commit='964effb4e883102d7c8cae627dbac4ba5d216a75')
+    version('1.4.0', commit='90a2eab66ff82ba6dd7fbb33e41cd0ded20fa218')
+
+    depends_on('r@3.3:', type=('build', 'run'))
```
The latest version doesn't have any specific R version requirement. Did it used to?

codeandkey replied:
Seems all the versions before bioconductor 3.9 did strangely, fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-variantannotation/package.py:
```diff
@@ -7,28 +7,53 @@
+    depends_on('r-rsamtools@1.99.0:', when='@1.30.1:', type=('build', 'run'))
+    depends_on('r-rhtslib', when='@1.30.1:', type=('build', 'run'))
+
+    depends_on('gmake')
```
```suggestion
    depends_on('gmake', type='build')
```

codeandkey replied:
corrected!

adamjstewart commented:
It looks like there are still several requested changes that haven't been resolved. I don't have the time to go through them again, but I'm trusting you to resolve all of them. Most of the packages in this PR have new releases now, and the dependencies have changed, making it hard to review. Which is why I want to get it merged ASAP so someone else can work out the bugs if they need to. Maybe that can be done in the "add a BundlePackage to R-Bioconductor" PR since we'll be able to test everything.

aweits commented:
Having struggled with this set of packages somewhat recently: Don't things become easier to deal with if you pull the branch tag associated with the bioconductor release? Dealing with the individual package versions seems almost counterproductive (not to mention confusing to end users, who are more likely interested in loading a bioconductor version)
```
coltrane (GIT/BiocFileCache):git branch -a
* master
  remotes/origin/HEAD -> origin/master
  remotes/origin/RELEASE_3_10
  remotes/origin/RELEASE_3_5
  remotes/origin/RELEASE_3_6
  remotes/origin/RELEASE_3_7
  remotes/origin/RELEASE_3_8
  remotes/origin/RELEASE_3_9
  remotes/origin/master
```

adamjstewart commented:
@aweits the plan is to create a `r-bioconductor` `BundlePackage` so that users can install a particular release without having to know which versions they need.

aweits commented:
@adamjstewart Understood. I guess the magic is going to be in the tool that generates the packages - will that code be available?

codeandkey commented:
@aweits The source code for the tool used is available [here](https://github.com/codeandkey/spack-biotool). It uses the release branches like you described, pulling from RELEASE_3_5 through RELEASE_3_5.

adamjstewart commented on var/spack/repos/builtin/packages/r-rhtslib/package.py:
```diff
@@ -7,17 +7,26 @@
+
+    depends_on('bzip2', type=('build', 'run'))
+    depends_on('xz', type=('build', 'run'))
+    depends_on('curl', type=('build', 'run'))
```
It doesn't link to any of these libraries? You can check by running `ldd -r` (on Linux) or `otool -L` (on macOS) on the shared object libraries or executables.

codeandkey replied:
I went ahead and added them as link dependencies too.

codeandkey commented:
@adamjstewart Sorry about the unfixed problems from the last review; I think I rebased the branch improperly and undid my own changes. I think I've addressed all the suggestions from the last review now.

adamjstewart commented on var/spack/repos/builtin/packages/r-affy/package.py:
```diff
@@ -7,19 +7,27 @@
+    depends_on('r-biocgenerics@0.1.12:', type=('build', 'run'))
+    depends_on('r-biobase@2.5.5:', type=('build', 'run'))
+    depends_on('r-affyio@1.13.3:', type=('build', 'run'))
     depends_on('r-biocinstaller', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-ampliqueso/package.py:
```diff
@@ -7,26 +7,32 @@
+       The package provides tools and reports for the analysis of amplicon
+       sequencing panels, such as AmpliSeq"""
+
+    homepage = "https://bioconductor.org/packages/ampliQueso"
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-biostrings/package.py:
```diff
@@ -7,23 +7,33 @@
+    depends_on('r-xvector@0.11.6:', type=('build', 'run'))
+
+    depends_on('r-s4vectors@0.17.25:', when='@2.48.0:', type=('build', 'run'))
+    depends_on('r-iranges@2.13.24:', when='@2.48.0:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-clusterprofiler/package.py:
```diff
@@ -7,22 +7,35 @@
-    depends_on('r-dose', type=('build', 'run'))
+    depends_on('r-ggplot2', type=('build', 'run'))
+    depends_on('r-go-db', type=('build', 'run'))
+    depends_on('r-gosemsim@2.0.0:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-cner/package.py:
```diff
@@ -32,6 +39,8 @@ class RCner(RPackage):
     depends_on('r-powerlaw@0.60.3:', type=('build', 'run'))
     depends_on('r-annotate@1.50.0:', type=('build', 'run'))
     depends_on('r-go-db@3.3.0:', type=('build', 'run'))
+    depends_on('r-utils@2.3.0:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-complexheatmap/package.py:
```diff
@@ -7,20 +7,34 @@
-    depends_on('r-dendextend', type=('build', 'run'))
-    depends_on('r-globaloptions', type=('build', 'run'))
-    depends_on('r@3.4.0:3.4.9', when='@1.14.0')
+    depends_on('r-dendextend@1.0.1:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed!

adamjstewart commented on var/spack/repos/builtin/packages/r-dose/package.py:
```diff
@@ -7,27 +7,37 @@
+    depends_on('r-fgsea', type=('build', 'run'))
+    depends_on('r-ggplot2', type=('build', 'run'))
+    depends_on('r-gosemsim@2.0.0:', type=('build', 'run'))
+    depends_on('r-igraph', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-dose/package.py:
```diff
@@ -7,27 +7,37 @@
+    depends_on('r-qvalue', type=('build', 'run'))
+    depends_on('r-reshape2', type=('build', 'run'))
+    depends_on('r-s4vectors', type=('build', 'run'))
+    depends_on('r-scales', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-ggraph/package.py:
```diff
@@ -0,0 +1,21 @@
+    version('1.0.2', sha256='4c24739ecabfc65c290a11980491a20bdaac675392a98dc30ccde76ac4b8f53a')
+
+    depends_on('r@2.10:', type=('build', 'run'))
+    depends_on('r-ggplot2@2.0.0:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed, replaced with newer recipe

adamjstewart commented on var/spack/repos/builtin/packages/r-glimma/package.py:
```diff
@@ -7,18 +7,28 @@
 
-    depends_on('r@3.5.0:3.5.9', when='@1.8.2:', type=('build', 'run'))
+    depends_on('r@3.3.0:', type=('build', 'run'))
+    depends_on('r-biobase', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-glimma/package.py:
```diff
@@ -7,18 +7,28 @@
+    depends_on('r-biobase', type=('build', 'run'))
     depends_on('r-edger', type=('build', 'run'))
     depends_on('r-jsonlite', type=('build', 'run'))
+    depends_on('r-scater', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-gostats/package.py:
```diff
@@ -7,21 +7,30 @@
+    depends_on('r@2.10:', type=('build', 'run'))
+    depends_on('r-biobase@1.15.29:', type=('build', 'run'))
+    depends_on('r-category@2.3.26:', type=('build', 'run'))
+    depends_on('r-graph@1.15.15:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-gtrellis/package.py:
```diff
@@ -7,20 +7,28 @@
     depends_on('r-getoptlong', type=('build', 'run'))
-    depends_on('r@3.4.0:3.4.9', when='@1.8.0')
+
+    depends_on('r-circlize@0.4.8:', when='@1.16.1:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed

adamjstewart commented on var/spack/repos/builtin/packages/r-hdf5array/package.py:
```diff
@@ -0,0 +1,47 @@
+
+    depends_on('r-delayedarray@0.9.3:', when='@1.12.3:', type=('build', 'run'))
+    depends_on('r-s4vectors@0.21.6:', when='@1.12.3:', type=('build', 'run'))
+    depends_on('r-rhdf5lib', when='@1.12.3:', type=('build', 'run'))
```
Ping

codeandkey replied:
fixed!

adamjstewart commented:
There's 50+ unresolved comments that I pinged, they might be hidden on the main view but if you go to the file diff you can see them.

adamjstewart commented:
Only 7 more packages with comments! Also, one of the unit tests is failing due to an invalid character in one of your packages. You should be able to grep for it.

adamjstewart commented on var/spack/repos/builtin/packages/r-oligoclasses/package.py:
```diff
@@ -7,25 +7,36 @@
     depends_on('r-foreach', type=('build', 'run'))
-    depends_on('r-biocinstaller', type=('build', 'run'))
-    depends_on('r-s4vectors', type=('build', 'run'))
+    depends_on('r-biocinstaller', when='1.38.0:1.42.0', type=('build', 'run'))
```
```suggestion
    depends_on('r-biocinstaller', when='@1.38.0:1.42.0', type=('build', 'run'))
```

codeandkey replied:
Nice catch! Committed.

adamjstewart commented:
Thanks @codeandkey! Now let's never do this again. Next time, I want 186 separate PRs 😝
